File: frontend/lib/api/firmware.ts
// Typed client for the daemon's `/api/firmware/*` HTTP surface (SC-10).
//
// Same bridge-then-env-port resolution as `./updater.ts`. Mirrors the
// route shapes documented in `backend/src/http/routes/firmware.rs` and
// the `FirmwareUpdater::State` enum in `backend/src/firmware/mod.rs` —
// keep this file in lockstep with those.
//
// Flash endpoints are intentionally implemented here against the
// SC-10-shipped 501 stubs so the UI can probe them and render a
// "Coming in SC-13" affordance. They will start returning real
// success / failure shapes once SC-13 lands.

import { getBridge } from "./client";

export type FirmwareStateKind =
  | "idle"
  | "up_to_date"
  | "available"
  | "downloading"
  | "ready"
  | "flashing"
  | "failed";

/**
 * Coarse phase the loader subprocess is currently in. The daemon
 * advances it by pattern-matching teensy_loader_cli stdout — see
 * `backend/src/firmware/flash.rs`. The UI stepper modal maps each
 * value onto a step screen.
 */
export type FlashPhase =
  | "starting"
  | "waiting_for_device"
  | "programming"
  | "booting";

export interface FirmwareState {
  kind: FirmwareStateKind;
  installed?: string | null;
  latest?: string;
  channel?: "stable" | "nightly";
  notes_url?: string | null;
  asset_url?: string;
  asset_name?: string;
  asset_size?: number;
  checked_at?: string;
  bytes_so_far?: number;
  total_bytes?: number | null;
  percent?: number | null;
  hex_path?: string;
  size?: number;
  sha256?: string;
  // Present only when kind === "flashing".
  version?: string;
  started_at?: string;
  /** Updates restructure: drives the stepper modal step. */
  phase?: FlashPhase;
  /**
   * Updates restructure: last ~20 stdout/stderr lines from the loader.
   * Capped daemon-side at `LOG_TAIL_CAP`. Used by the stepper modal
   * to render recent loader output in a muted mono block.
   */
  log_tail?: string[];
  error?: string;
  when?: string;
}

export interface FirmwareStatusResponse {
  state: FirmwareState;
  installed_version: string | null;
  channel: "stable" | "nightly" | "unknown";
  repo: string;
  board: string | null;
  auto_check: boolean;
  experimental_builds: boolean;
  /**
   * True when the bundled `teensy_loader_cli.exe` is where the daemon
   * expects it (set by electron via STREAMCHEATS_TEENSY_LOADER_PATH).
   * With a correct install this is always true; the UI surfaces a
   * "Flash tool missing — please reinstall" message when it isn't.
   */
  loader_ready: boolean;
}

export interface FirmwareReleaseEntry {
  version: string;
  channel: "stable" | "nightly";
  commit: string | null;
  board: string;
  published_at: string | null;
  asset_url: string;
  asset_name: string;
  asset_size: number;
  html_url: string | null;
}

export interface FirmwareReleasesResponse {
  releases: FirmwareReleaseEntry[];
}

/**
 * Result of dispatching a flash. 202 on accepted, 409 with a stable
 * error code on rejection (see backend/src/firmware/mod.rs:
 * `start_flash` / `start_flash_local`). The UI polls
 * `/api/firmware/status` for progress once dispatch succeeds.
 */
export type FlashResult =
  | { ok: true }
  | {
      ok: false;
      reason:
        | "flash_in_progress"
        | "hex_not_downloaded"
        | "unknown_version"
        | "unsupported_board"
        | "invalid_hex"
        // Daemon couldn't resolve the bundled loader binary. With a
        // correct install this never fires; the UI surfaces a
        // "please reinstall" message when it does.
        | "loader_unavailable"
        | "network"
        // Kept for legacy paths (e.g. an older daemon still returning
        // 501); current daemon never produces this.
        | "not_implemented"
        | "unknown";
      detail?: string;
    };

async function resolveBase(): Promise<string | null> {
  const bridge = getBridge();
  if (bridge?.getBackendUrl) {
    const r = await bridge.getBackendUrl();
    if (r.ok) return r.http;
  }
  const port = process.env.NEXT_PUBLIC_STREAMCHEATS_HTTP_PORT;
  if (port) return `http://127.0.0.1:${port}`;
  return null;
}

async function call<T = unknown>(
  path: string,
  init?: RequestInit
): Promise<T | null> {
  const base = await resolveBase();
  if (!base) return null;
  try {
    const resp = await fetch(`${base}${path}`, init);
    // 202 (accepted) and 409 (conflict) still carry JSON bodies that
    // the caller wants. 501 (flash stubs) is handled separately by
    // `flash` / `flashLocal` below.
    if (!resp.ok && resp.status !== 202 && resp.status !== 409) {
      return null;
    }
    return (await resp.json()) as T;
  } catch {
    return null;
  }
}

export function getStatus(): Promise<FirmwareStatusResponse | null> {
  return call<FirmwareStatusResponse>("/api/firmware/status");
}

export function getReleases(): Promise<FirmwareReleasesResponse | null> {
  return call<FirmwareReleasesResponse>("/api/firmware/releases");
}

export function checkNow(): Promise<{ state: FirmwareState } | null> {
  return call<{ state: FirmwareState }>("/api/firmware/check", {
    method: "POST",
  });
}

export function startDownload(
  version: string
): Promise<{ ok: boolean; error?: string } | null> {
  return call("/api/firmware/download", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ version }),
  });
}

/**
 * Probe `/api/firmware/flash`. Returns `{ ok: false, reason:
 * "not_implemented" }` until SC-13 wires up `teensy_loader_cli`.
 */
export async function flash(version: string): Promise<FlashResult> {
  const base = await resolveBase();
  if (!base) return { ok: false, reason: "network" };
  try {
    const resp = await fetch(`${base}/api/firmware/flash`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ version }),
    });
    return await interpretFlashResponse(resp);
  } catch (e) {
    return {
      ok: false,
      reason: "network",
      detail: e instanceof Error ? e.message : undefined,
    };
  }
}

export async function flashLocal(hexPath: string): Promise<FlashResult> {
  const base = await resolveBase();
  if (!base) return { ok: false, reason: "network" };
  try {
    const resp = await fetch(`${base}/api/firmware/flash_local`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ hex_path: hexPath }),
    });
    return await interpretFlashResponse(resp);
  } catch (e) {
    return {
      ok: false,
      reason: "network",
      detail: e instanceof Error ? e.message : undefined,
    };
  }
}

async function interpretFlashResponse(resp: Response): Promise<FlashResult> {
  if (resp.status === 501) {
    return { ok: false, reason: "not_implemented" };
  }
  if (resp.status === 202) {
    return { ok: true };
  }
  // Daemon returns 503 when the bundled loader binary can't be
  // resolved (broken install). The UI surfaces a "please reinstall"
  // message rather than presenting it as a transient conflict.
  if (resp.status === 503) {
    try {
      const body = (await resp.json()) as { error?: string };
      if (body.error === "loader_unavailable") {
        return { ok: false, reason: "loader_unavailable" };
      }
      return { ok: false, reason: "unknown", detail: body.error };
    } catch {
      return { ok: false, reason: "loader_unavailable" };
    }
  }
  if (resp.status === 409) {
    try {
      const body = (await resp.json()) as { error?: string };
      const raw = body.error ?? "";
      // The daemon returns a stable code; for "invalid_hex" it prefixes
      // a human-readable cause (e.g. "invalid_hex: ... is empty").
      if (raw.startsWith("invalid_hex")) {
        return {
          ok: false,
          reason: "invalid_hex",
          detail: raw.replace(/^invalid_hex:\s*/, ""),
        };
      }
      if (
        raw === "flash_in_progress" ||
        raw === "hex_not_downloaded" ||
        raw === "unknown_version" ||
        raw === "unsupported_board"
      ) {
        return { ok: false, reason: raw };
      }
      return { ok: false, reason: "unknown", detail: raw };
    } catch {
      return { ok: false, reason: "unknown" };
    }
  }
  return { ok: false, reason: "unknown", detail: `HTTP ${resp.status}` };
}

/**
 * Result of `POST /api/firmware/cancel_flash`. The daemon returns 202
 * `{ ok: true }` when it signalled the supervision loop to kill the
 * subprocess (subsequent state polls will show `Failed { error:
 * "user_cancelled" }`), or 409 `{ ok: false, reason: "not_flashing" }`
 * when nothing was in flight.
 */
export type CancelFlashResult =
  | { ok: true }
  | {
      ok: false;
      reason: "not_flashing" | "network" | "unknown";
      detail?: string;
    };

/**
 * Updates restructure: POST `/api/firmware/cancel_flash`. Used by the
 * stepper modal's Cancel button in the `waiting_for_device` phase
 * (NOT in `programming` — interrupting mid-write would brick the
 * device). The daemon kills the subprocess and transitions to
 * `Failed { error: "user_cancelled" }`; the polling status watcher
 * picks the transition up on the next tick.
 */
export async function cancelFlash(): Promise<CancelFlashResult> {
  const base = await resolveBase();
  if (!base) return { ok: false, reason: "network" };
  try {
    const resp = await fetch(`${base}/api/firmware/cancel_flash`, {
      method: "POST",
    });
    if (resp.status === 202) return { ok: true };
    if (resp.status === 409) {
      try {
        const body = (await resp.json()) as { error?: string };
        if (body.error === "not_flashing") {
          return { ok: false, reason: "not_flashing" };
        }
        return { ok: false, reason: "unknown", detail: body.error };
      } catch {
        return { ok: false, reason: "unknown" };
      }
    }
    return { ok: false, reason: "unknown", detail: `HTTP ${resp.status}` };
  } catch (e) {
    return {
      ok: false,
      reason: "network",
      detail: e instanceof Error ? e.message : undefined,
    };
  }
}

/** Resolved shape returned by `window.streamcheats.pickHexFile`. */
export type PickHexFileResult =
  | { ok: true; path: string }
  | { ok: false; reason: "cancelled" | "unavailable" };

/**
 * Open the OS-native file picker (constrained to `.hex`) via the
 * Electron preload bridge. Returns the absolute path the user chose,
 * or null on cancel / no-bridge.
 */
export async function pickHexFile(): Promise<string | null> {
  const bridge = getBridge();
  if (!bridge || typeof bridge.pickHexFile !== "function") return null;
  const r = await bridge.pickHexFile();
  if (r.ok) return r.path;
  return null;
}

File: backend/src/http/routes/firmware.rs
//! `/api/firmware/*` routes — see SC-10 + SC-13.
//!
//! Read-side endpoints (`status`, `releases`) snapshot the firmware
//! updater's shared state. Write-side endpoints (`check`, `download`,
//! `flash`, `flash_local`) kick off background work and return 202 on
//! dispatch.
//!
//! Flash dispatch (SC-13): both `/flash` and `/flash_local` first
//! validate single-flight + path/version preconditions and return
//! 409 with a stable error code on rejection; otherwise they spawn
//! the background `teensy_loader_cli` task and return 202. If the
//! daemon can't find the `teensy_loader_cli.exe` binary at all (not
//! bundled in this build), the failure is reported via the state
//! machine's `Failed` transition with `binary_not_bundled`-shaped
//! text — the synchronous response is still 202 because resolving
//! the binary is a runtime concern inside the spawn.

use axum::extract::State;
use axum::http::StatusCode;
use axum::Json;
use serde::Deserialize;
use serde_json::json;
use tracing::{info, warn};

use crate::firmware::device::InstalledFirmware;
use crate::http::state::AppState;


/// `GET /api/firmware/status` — installed version (from heartbeat) +
/// the current updater [`crate::firmware::State`] + the configured
/// repo + the board reported by the most-recently-seen release asset
/// (so the UI doesn't hard-code `teensy-4.1`).
pub async fn status(State(state): State<AppState>) -> Json<serde_json::Value> {
    let snap = state.firmware.snapshot().await;
    let releases = state.firmware.releases().await;
    let installed = state.firmware.installed.snapshot();

    let (installed_version, installed_channel) = match installed {
        InstalledFirmware::Known { version, .. } => (Some(format!("rel-{}", version)), "unknown"),
        InstalledFirmware::Unknown => (None, "unknown"),
    };
    let board = releases.first().map(|r| r.board.clone());

    // Cheap synchronous "is the bundled loader where we expect it?"
    // check. We do NOT shell out to `--help` on every status poll —
    // that's reserved for the resolve path inside `start_flash`.
    // With a correct install this is always true; surface it so the UI
    // can show a clear "please reinstall" error if it ever isn't.
    let loader_ready = state.firmware.loader_present();

    Json(json!({
        "state": snap,
        "installed_version": installed_version,
        // Channel of the installed firmware. Without the commit suffix
        // the device-reported version can't disambiguate stable vs.
        // nightly of the same major.minor — surface `"unknown"` until
        // the firmware exposes it explicitly (then SC-13 can refine).
        "channel": installed_channel,
        "repo": state.firmware.repo().await,
        "board": board,
        "auto_check": state.firmware.auto_check(),
        "experimental_builds": state.firmware.experimental(),
        // Existence-only check for the bundled loader. With a correct
        // install this is always true; the UI surfaces "Flash tool
        // missing — please reinstall" when it isn't.
        "loader_ready": loader_ready,
    }))
}

/// `GET /api/firmware/releases` — full list of releases (sorted newest
/// first by `published_at`). Filtering by channel / search is done by
/// the UI — the assumption is the list stays small enough not to need
/// server-side pagination for v1.
pub async fn releases(State(state): State<AppState>) -> Json<serde_json::Value> {
    let entries = state.firmware.releases().await;
    Json(json!({ "releases": entries }))
}

/// `POST /api/firmware/check` — kick off an immediate check. Returns
/// the new status snapshot.
pub async fn check(State(state): State<AppState>) -> Json<serde_json::Value> {
    state.firmware.check_once().await;
    let snap = state.firmware.snapshot().await;
    Json(json!({ "state": snap }))
}

#[derive(Deserialize)]
pub struct DownloadBody {
    pub version: String,
}

/// `POST /api/firmware/download` — start downloading a specific
/// release. Body: `{ "version": "rel-5.17" }`.
pub async fn download(
    State(state): State<AppState>,
    Json(body): Json<DownloadBody>,
) -> (StatusCode, Json<serde_json::Value>) {
    match state.firmware.start_download(&body.version).await {
        Ok(()) => {
            info!("firmware: download dispatched for {}", body.version);
            (StatusCode::ACCEPTED, Json(json!({ "ok": true })))
        }
        Err(e) => {
            warn!("firmware: download rejected: {}", e);
            (
                StatusCode::CONFLICT,
                Json(json!({ "ok": false, "error": e })),
            )
        }
    }
}

#[derive(Deserialize)]
pub struct FlashBody {
    pub version: String,
}

/// `POST /api/firmware/flash` — flash a specific previously-downloaded
/// release. Body: `{ "version": "rel-5.17" }`. The release must be in
/// `Ready` state for that exact version (i.e. user already hit
/// Download). Returns 202 on dispatch; 409 with `{ error: "..." }` on
/// rejection. Error codes mirror [`crate::firmware::FirmwareUpdater::start_flash`].
pub async fn flash(
    State(state): State<AppState>,
    Json(body): Json<FlashBody>,
) -> (StatusCode, Json<serde_json::Value>) {
    match state.firmware.start_flash(&body.version).await {
        Ok(()) => {
            info!("firmware: flash dispatched for {}", body.version);
            (StatusCode::ACCEPTED, Json(json!({ "ok": true })))
        }
        Err(e) => {
            warn!("firmware: flash rejected: {}", e);
            // Surface loader-not-available as 503 so the UI can show a
            // "please reinstall" message rather than treating it like a
            // transient conflict.
            let status = if e == "loader_unavailable" {
                StatusCode::SERVICE_UNAVAILABLE
            } else {
                StatusCode::CONFLICT
            };
            (status, Json(json!({ "ok": false, "error": e })))
        }
    }
}

#[derive(Deserialize)]
pub struct FlashLocalBody {
    pub hex_path: String,
}

/// `POST /api/firmware/flash_local` — flash an arbitrary local `.hex`
/// file. Body: `{ "hex_path": "C:\\absolute\\path\\to\\firmware.hex" }`.
/// Useful for downgrading to an older firmware not in the release
/// feed. Validation: file exists, `.hex` extension, non-empty. Same
/// single-flight semantics as `/flash`.
pub async fn flash_local(
    State(state): State<AppState>,
    Json(body): Json<FlashLocalBody>,
) -> (StatusCode, Json<serde_json::Value>) {
    let path = std::path::PathBuf::from(&body.hex_path);
    match state.firmware.start_flash_local(path).await {
        Ok(()) => {
            info!("firmware: flash_local dispatched for {}", body.hex_path);
            (StatusCode::ACCEPTED, Json(json!({ "ok": true })))
        }
        Err(e) => {
            warn!("firmware: flash_local rejected: {}", e);
            let status = if e == "loader_unavailable" {
                StatusCode::SERVICE_UNAVAILABLE
            } else {
                StatusCode::CONFLICT
            };
            (status, Json(json!({ "ok": false, "error": e })))
        }
    }
}

/// `POST /api/firmware/cancel_flash` — kill the in-flight flash. Returns
/// 202 `{ ok: true }` on dispatch (the supervision loop in
/// [`crate::firmware::flash::run_flash`] will kill the subprocess and
/// transition the state machine to `Failed { error: "user_cancelled", ... }`
/// shortly after). Returns 409 `{ ok: false, error: "not_flashing" }`
/// when nothing is in flight, so a stray button press doesn't silently
/// no-op.
pub async fn cancel_flash(
    State(state): State<AppState>,
) -> (StatusCode, Json<serde_json::Value>) {
    if state.firmware.cancel_flash().await {
        info!("firmware: cancel_flash signalled");
        (StatusCode::ACCEPTED, Json(json!({ "ok": true })))
    } else {
        (
            StatusCode::CONFLICT,
            Json(json!({ "ok": false, "error": "not_flashing" })),
        )
    }
}
